### ISSUE TITLE
Question About RL and Fine-Tuning Runtime (H100 vs A100)

### ISSUE THREAD

**Issue by AngusHsuPhys** [schwallergroup/matinvent#1]
Dear developers,

I’m writing to ask about the estimated execution time for the example workflow you provided. In the paper, the RL and fine-tuning steps are reported to have been run on an NVIDIA H100 GPU, and I was hoping to get a rough sense of the corresponding wall-clock time.

For reference, I’m currently running the same loop on a newly constructed cluster using a single NVIDIA A100 (40 GB). The total runtime is approximately 80,666 seconds (~22.4 hours).

Could you share the approximate runtime you observed when running this on an H100, or any expected speedup relative to an A100? This information would be very helpful for benchmarking and experiment planning. Additionally, I’d appreciate any suggestions on settings or configurations that could help optimize performance.

Thanks in advance for your help!

**Comment by jwchen25** [comment]
Could you tell me your experimental parameters, such as which diffusion model, which reward, and the maximum number of RL iterations?

**Comment by jwchen25** [comment]
When I run the following task on a single H100 GPU, the total time is about 53,049 seconds (14.7 hours).

```bash
python -u main.py \
    pipeline=mat_invent \
    model=mattergen \
    reward=hhi
```

The main parameters are: rl_epoch=120, eval_size=16

**Comment by AngusHsuPhys** [comment]
Sorry for the late reply — thanks for sharing the total runtime for comparison; I really appreciate it.

I was following the `default` settings, which should be the exact same parameter as you used. 
```
python -u main.py \
    expname=test \
    pipeline=mat_invent \
    model=mattergen \
    reward=hhi \
    logger=wandb
```
with the same paramters for `rl_epoch=120`, `eval_size=16`.

I’m a bit surprised the gap is this large. Do you know if there are any other factors that could affect throughput (e.g., precision settings, number of CPU workers, or environment/reward evaluation parallelism)?

**Comment by jwchen25** [comment]
Typically, Loop 0 takes longer because it requires loading many things (such as a reference dataset for calculating the ehull and SUN ratios). On a machine with 64-core CPU and a single H100 GPU, Loop 0 takes approximately 12 minutes. Each subsequent loop takes about 7 minutes. You can check the time taken for each step in a loop to determine the steps that limit the rate. Based on my results, the most time-consuming steps were sampling (~2 min) and fine-tuning (~4 min). You can improve the speed by modifying the parameters:

- reduce the sampling number [**sample_cfg.batch_size**](https://github.com/schwallergroup/matinvent/blob/main/configs/model/mattergen.yaml#L7), for example, change from `${calc:'${eval_size} * 4'}` to `${calc:'${eval_size} * 3'}`;
-  increase the [**accum_steps**](https://github.com/schwallergroup/matinvent/blob/main/configs/pipeline/mat_invent.yaml#L32) during finetuning, for example, change from 50 to 100 or 200;
- decrease [**epochs**](https://github.com/schwallergroup/matinvent/blob/main/configs/pipeline/mat_invent.yaml#L33) for fine-tuning, for example, change from 3 to 2;

**It is worth noting that modifying these parameters will improve running speed, but will affect the performance of RL.**

Below are my results:
```
[2025-07-14 13:40:27,789][root][INFO] - *****   LOOP 1 START   *****
[2025-07-14 13:40:27,789][root][INFO] - SAMPLE:
^MGenerating samples:   0%|          | 0/1 [00:00<?, ?it/s]
^M  0%|          | 0/1000 [00:00<?, ?it/s]^[[A
^M  5%|▌         | 50/1000 [00:06<02:01,  7.80it/s]^[[A
...(simplify)...
^M100%|██████████| 1000/1000 [02:05<00:00,  8.10it/s]^[[A
[2025-07-14 13:42:37.149[root][INFO][mattersim.forcefield.potential] [Loading the pre-trained mattersim-v1.0.0-5M.pth model
[2025-07-14 13:43:14,520][root][INFO] - Number of filtered samples: 17
[2025-07-14 13:43:14,521][root][INFO] - avg_energy_above_hull_per_atom: 0.063197, avg_rmsd_from_relaxation: 0.054379, frac_novel_unique_stable_structures: 0.303571, frac_stable_structures: 0.821429, frac_successful_jobs: 1.000000, avg_comp_validity: 1.000000, avg_structure_comp_validity: 1.000000, avg_structure_validity: 1.000000, frac_novel_structures: 0.482143, frac_novel_systems: 0.089286, frac_novel_unique_structures: 0.482143, frac_unique_structures: 1.000000, frac_unique_systems: 1.000000, precision: 0.517857, recall: 0.000112
[2025-07-14 13:43:14,528][root][INFO] - SCORE:
[2025-07-14 13:43:14,534][root][INFO] - Evaluation costs to date: 32
[2025-07-14 13:43:14,534][root][INFO] - Number of samples that successfully obtained rewards: 16
[2025-07-14 13:43:14,534][root][INFO] - reward mean=0.1296 std=0.1949
[2025-07-14 13:43:14,534][root][INFO] - hhi mean=3620.8718 std=1204.3451
[2025-07-14 13:43:14,586][root][INFO] - 32 crystals generated so far, 32 unique components.  Burden: None, Div. Ratio: 1.0.
[2025-07-14 13:43:14,590][root][INFO] - Diversity filter: tol_n=0, buff_n=0
[2025-07-14 13:43:14,592][root][INFO] - replay buffer size=14
[2025-07-14 13:43:14,593][root][INFO] - buffer reward mean=0.3779154475286561
[2025-07-14 13:43:14,593][root][INFO] - FINETUNE:
[2025-07-14 13:44:30,972][root][INFO] - Epoch 0: loss: 0.1338, loss_diff: 0.1300, loss_kl: 0.1501
[2025-07-14 13:45:46,729][root][INFO] - Epoch 1: loss: 0.1318, loss_diff: 0.1274, loss_kl: 0.1740
[2025-07-14 13:47:02,333][root][INFO] - Epoch 2: loss: 0.1305, loss_diff: 0.1247, loss_kl: 0.2344
[2025-07-14 13:47:02,335][root][INFO] - *****   LOOP 1 FINISH   *****
[2025-07-14 13:47:02,335][root][INFO] - Total time taken: 6.58 min.
```
